include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+   * @param details are details of any (gRPC) error.
+   */
+  virtual void sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                                 Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                                 string_view /* details */) {
+    unimplemented();
+  }
+
+  // Shared Data
+
+  /**
+   * Get proxy-wide key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a number which will be incremented when a data value has been changed.
+   * @param data is a location to store the returned stored 'value' and the corresponding 'cas'
+   * compare-and-swap value which can be used with setSharedData for safe concurrent updates.
+   */
+  virtual WasmResult
+  getSharedData(string_view /* key */,
+                std::pair<std::string /* value */, uint32_t /* cas */> * /* data */);
+
+  /**
+   * Set a key-value data shared between VMs.
+   * @param key is a proxy-wide key mapping to the shared data value.
+   * @param cas is a compare-and-swap value. If it is zero it is ignored, otherwise it must match
+   * the cas associated with the value.
+   * @param data is a location to store the returned value.
+   */
+  virtual WasmResult setSharedData(string_view /* key */, string_view /* value */,
+                                   uint32_t /* cas */);
+
+  // Shared Queue
+
+  /**
+   * Register a proxy-wide queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult registerSharedQueue(string_view /* queue_name */, uint32_t *token_ptr);
+
+  /**
+   * Get the token for a queue.
+   * @param vm_id is the vm_id of the Plugin of the Root Context which registered the queue.
+   * @param queue_name is a name for the queue. The queue_name is combined with the vm_id (if any)
+   * to make a unique identifier for the queue.
+   * @param token_ptr a location to store a token corresponding to the queue.
+   */
+  virtual WasmResult resolveSharedQueue(string_view /* vm_id */, string_view /* queue_name */,
+                                        uint32_t * /* token_ptr */);
+
+  /**
+   * Dequeue a message from a shared queue.
+   * @param token is a token returned by registerSharedQueue();
+   * @param data_ptr is a location to store the data dequeued.
+   */
+  virtual WasmResult dequeueSharedQueue(uint32_t /* token */, std::string * /* data_ptr */);
+
+  /**
+   * Enqueue a message on a shared queue.
+   * @param token is a token returned by resolveSharedQueue();
+   * @param data is the data to be queued.
+   */
+  virtual WasmResult enqueueSharedQueue(uint32_t /* token */, string_view /* value */);
+
+  // Header/Trailer/Metadata Maps
+
+  /**
+   * Add a key-value pair to a header map.
```

Comment:
I think you have to be super-precise here with semantics on header map, as they keep changing in internally in Envoy and put the WASM API stability at risk. For example, what happens when you add multiple values for the same key? Is there concatenation?

Reply:
I agree.  That should be laid out in the spec https://github.com/proxy-wasm/spec see pull/1 to comment.  Then we can copy that information here. 

Personally I would like to see the ABI handling multiple values rather than punting to the plugin writer.